src/prommis/nanofiltration/diafiltration.py

```diff
@@ -0,0 +1,737 @@
+        )
+
+    return results
+
+
+def add_costing(m):
+    """
+    Method to add costing block to the flowsheet
+
+    Args:
+        m: Pyomo model
+    """
+    # creating dummy variables to store the UnitModelCostingBlocks
+    m.fs.membrane = UnitModelBlock()
+    m.fs.feed_pump = UnitModelBlock()
+    m.fs.diafiltrate_pump = UnitModelBlock()
+
+    m.fs.costing = DiafiltrationCosting()
+    m.fs.membrane.costing = UnitModelCostingBlock(
+        flowsheet_costing_block=m.fs.costing,
+        costing_method=DiafiltrationCostingData.cost_membranes,
+        costing_method_arguments={
+            "membrane_length": m.membrane_length,  # total membrane length
+            "membrane_width": w,  # membrane width
+            "water_flux": Jw,  # water flux
+            "vol_flow_feed": m.fs.stage3.retentate_side_stream_state[
+                0, 10
+            ].flow_vol,  # feed
+            "vol_flow_perm": m.fs.stage3.permeate_outlet.flow_vol[0],  # permeate
+        },
+    )
+    m.fs.feed_pump.costing = UnitModelCostingBlock(
+        flowsheet_costing_block=m.fs.costing,
+        costing_method=DiafiltrationCostingData.cost_pump,
+        costing_method_arguments={
+            "inlet_pressure": atmospheric_pressure
+            + units.convert(m.fs.membrane.costing.pressure_drop, to_units=units.Pa),
+            "outlet_pressure": 1e-5  # assume numerically 0 since SEC accounts for feed pump OPEX
+            * units.psi,  # this should make m.fs.feed_pump.costing.fixed_operating_cost ~0
+            "inlet_vol_flow": m.fs.stage3.retentate_side_stream_state[
+                0, 10
+            ].flow_vol,  # feed
+        },
+    )
+    m.fs.diafiltrate_pump.costing = UnitModelCostingBlock(
+        flowsheet_costing_block=m.fs.costing,
+        costing_method=DiafiltrationCostingData.cost_pump,
+        costing_method_arguments={
+            "inlet_pressure": atmospheric_pressure,
+            "outlet_pressure": operating_pressure,
+            "inlet_vol_flow": m.fs.stage3.retentate_inlet.flow_vol[0],  # diafiltrate
+        },
+    )
+    m.fs.costing.cost_process()
+
+
+def unfix_opt_variables(m):
+    """
+    Method to unfix variables for performing optimization with DOF>0
+
+    Args:
+        m: Pyomo model
+    """
+    m.fs.stage1.length.unfix()
+    m.fs.stage2.length.unfix()
+    m.fs.stage3.length.unfix()
+
+
+def add_product_constraints(m):
+    """
+    Method to add recovery/purity constraints to the flowsheet for performing optimization
+
+    Args:
+        m: Pyomo model
+    """
+
+    @m.Constraint()
+    def Co_recovery_constraint(m):
+        return m.Co_recovery >= 0.4
+
+    @m.Constraint()
+    def Li_recovery_constraint(m):
+        return m.Li_recovery >= 0.95
+
+    # @m.Constraint()
+    # def Li_purity_constraint(m):
+    #     return m.Li_purity >= 0.8
+
+    # @m.Constraint()
+    # def Co_purity_constraint(m):
+    #     return m.Co_purity >= 0.5
```

Comment:
Is there a reason these purity constraints shouldn't be enforced?

Reply:
I didn't use them for this particular example, but the user could choose to use them. I will update these to be optional constraints.